haystack/components/eval/statistical_evaluator.py

```diff
@@ -0,0 +1,131 @@
+from .preprocess import _preprocess_text
+
+
+@component
+class StatisticalEvaluator:
+    """
+    StatisticalEvaluator is a component that evaluates the performance of a model based on statistical metrics.
+    It's usually used in QA and Retrieval Augmented Generation (RAG) pipelines to evaluate the quality of the generated answers.
+
+    The supported metrics are:
+    - F1: Measures word overlap between predictions and labels.
+    - Exact Match: Measures the proportion of cases where prediction is identical to the expected label.
+    """
+
+    class Metric(Enum):
+        """
+        Supported metrics
+        """
+
+        F1 = "F1"
+        EM = "Exact Match"
+
+    def __init__(
+        self,
+        labels: List[str],
+        metric: Metric,
+        regexes_to_ignore: Optional[List[str]] = None,
+        ignore_case: bool = False,
+        ignore_punctuation: bool = False,
+        ignore_numbers: bool = False,
+    ):
+        """
+        Creates a new instance of StatisticalEvaluator.
+
+        :param labels: The list of expected answers.
+        :param metric: Metric to use for evaluation in this component. Supported metrics are F1 and Exact Match.
+        :param regexes_to_ignore: A list of regular expressions. If provided, it removes substrings
+            matching these regular expressions from both predictions and labels before comparison. Defaults to None.
+        :param ignore_case: If True, performs case-insensitive comparison. Defaults to False.
+        :param ignore_punctuation: If True, removes punctuation from both predictions and labels before
+            comparison. Defaults to False.
+        :param ignore_numbers: If True, removes numerical digits from both predictions and labels
+            before comparison. Defaults to False.
+        """
+        self._labels = labels
+        self._metric = metric
+        self._regexes_to_ignore = regexes_to_ignore
+        self._ignore_case = ignore_case
+        self._ignore_punctuation = ignore_punctuation
+        self._ignore_numbers = ignore_numbers
+
+        self._metric_function = {
+            StatisticalEvaluator.Metric.F1: self._f1,
+            StatisticalEvaluator.Metric.EM: self._exact_match,
+        }[self._metric]
+
+    def to_dict(self) -> Dict[str, Any]:
+        return default_to_dict(
+            self,
+            labels=self._labels,
+            metric=self._metric.value,
+            regexes_to_ignore=self._regexes_to_ignore,
+            ignore_case=self._ignore_case,
+            ignore_punctuation=self._ignore_punctuation,
+            ignore_numbers=self._ignore_numbers,
+        )
+
+    @classmethod
+    def from_dict(cls, data: Dict[str, Any]) -> "StatisticalEvaluator":
+        data["init_parameters"]["metric"] = StatisticalEvaluator.Metric(data["init_parameters"]["metric"])
+        return default_from_dict(cls, data)
+
+    @component.output_types(result=float)
+    def run(self, predictions: List[str]) -> Dict[str, Any]:
```

Comment:
Yeah I also noticed the comment about length of these predictions, why not add a check here for zero length and then we can omit checks in all metrics?

Reply:
As of now I moved only the metrics that we already had from the old API. In future PRs other will be added, I'm unsure whether those will return the same values as F1 and Exact Match if length is zero. That's the only reason I've done it like this.